### PR TITLE
Add lockbud task to CI

### LINKED ISSUE #6413: Get lockbud running again
## Description

A while ago the author of `lockbud` found a bunch of nice deadlocks in Lighthouse.

- https://github.com/BurtonQin/lockbud

We are still encountering deadlocks occasionally, see:

- https://github.com/sigp/lighthouse/pull/6412

It would be great if we could run `lockbud` on CI to detect them statically.

## Steps to resolve

Last time we tried `lockbud` it had bitrotted and was incapable of compiling Lighthouse.

It seems there have been updates in the last few weeks that might make it work with more recent compiler versions!

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

Closes #6413

## Proposed Changes

Add lockbud to CI

## Additional Info

Lockbud was pinned to 2024-05-20 Rust nightly. This conflicted with lazy_lock as it used to be an unstable feature. I've opened a PR here to pin it to the current latest Rust nightly (2024-10-05). I'm not sure if my changes have introduced any regressions, hopefully the author will review it soon and give some feedback.

https://github.com/BurtonQin/lockbud/pull/70

Currently the new CI step uses a docker image hosted on my account

Steps to generate the docker image:

- Checkout https://github.com/eserilev/lockbud `docker-instance` branch
- Run `docker build --platform linux/amd64 -t [namespace]/[repository]  -f Dockerfile .  `
- Push the resulting image to docker hub

### CODE REVIEW

**Comment by eserilev** [comment]
We're getting a hit on a potential double lock here

```
[
      {
        "DoubleLock": {
          "bug_kind": "DoubleLock",
          "possibility": "Possibly",
          "diagnosis": {
            "first_lock_type": "ParkingLotRead(inner::DepositUpdater)",
            "first_lock_span": "beacon_node/eth1/src/service.rs:552:9: 553:20 (#0)",
            "second_lock_type": "ParkingLotRead(inner::DepositUpdater)",
            "second_lock_span": "beacon_node/eth1/src/service.rs:535:29: 535:51 (#0)",
            "callchains": [
              [
                [
                  "beacon_node/eth1/src/service.rs:555:40: 555:65 (#0)"
                ]
              ]
            ]
          },
          "explanation": "The first lock is not released when acquiring the second lock"
        }
      }
    ]
```

It seems to be a false positive as these are read locks

**Comment by jimmygchen** [comment]
> beacon_node/eth1/src/service.rs:555

Oh nice catch! i think this could potentially deadlock because we try to acquire a 2nd lock on `self.deposits` in the same thread, so if another thread tries to acquire a write lock in between, then we get a deadlock.

File: .github/workflows/test-suite.yml [hunk 1]
```diff
@@ -54,6 +54,20 @@ jobs:
          done
          echo "skip_ci=$SKIP_CI" >> $GITHUB_OUTPUT
 
+  lockbud:
```
Comment (jimmygchen) [inline .github/workflows/test-suite.yml:57]: need to add `lockbud` job to the list of `needs` in `test-suite-success` job

File: .github/workflows/test-suite.yml [hunk 2]
```diff
@@ -54,6 +54,20 @@ jobs:
...
+    name: lockbud
+    runs-on: ubuntu-latest
+    container:
+      image: eserilev/lockbud:latest
```
Comment (dapplion) [inline .github/workflows/test-suite.yml]: Is this a blocker to merging?

Comment (eserilev) [reply .github/workflows/test-suite.yml]: I've updated to point to `sigmaprime/lockbud`

**Review by dapplion — APPROVED** [review]
Great to have lockbud on CI :)

**Comment by michaelsproul** [comment]
@mergify queue

**Pull request merged**